src/app/content/courses/merkle-trees-on-solana/introduction/en.mdx

```diff
@@ -0,0 +1,159 @@
+
+3. Deal with odd numbers of nodes: it might happens that the number of nodes at any level is odd, making it "impossible" to pair it with another leaf. We can solve this by duplicating the last element to make pairs or by adding empty/null nodes with predefined hash values.
+
+### Merkle Proofs
+
+A Merkle proof consists of the minimal set of hashes needed to reconstruct the path from any leaf to the root.
+
+For leaf position `i` in a tree with `n` leaves, a proof requires exactly `⌊log₂(n)⌋` sibling hashes. This logarithmic proof size makes verification scalable even for massive datasets.
+
+To create a proof for data block at position `i`:
+
+1. Start at leaf position `i`
+
+2. For each level moving toward the root: identify the sibling node (adjacent hash at the same level), add the sibling hash to the proof and move to parent position `⌊current_position/2⌋`
+
+3. Repeat until reaching the root
+
+For a tree like this: 
+
+```
+        Root Hash (R)
+       /             \
+    H(A,B)          H(C,D)
+   /     \         /     \
+  H(A)   H(B)   H(C)   H(D)
+   |      |      |      |
+   A      B      C      D
+```
+
+The proof for the leaf C would look like this: `[H(D), H(AB)]`
+
+Based on the proof we can then verify that the leaf C exists in a tree with trusted root hash R like this:
+
+1. Start with the data: Compute `H(C)` as the starting point
+
+2. Apply the proof: For each sibling hash in the proof: determine if the current hash is a left (even index) or right (odd index) child, concatenate appropriately: `H(current || sibling)` for left hashes or `H(sibling || current)` for right hashes
+
+3. Compare roots: Hash the result till you get to the root and check if the final computed hash must equal the trusted root R
+
+For a tree like this: 
+
+```
+        Root Hash (R)
+       /             \
+    H(A,B)          H(C,D)
+   /     \         /     \
+  H(A)   H(B)   H(C)   H(D)
+   |      |      |      |
+   A      B      C      D
+```
+
+The verification for the leaf C would look like this: 
+1. Reconstruct the immediate parent: `H(C,D) = H(H(C), H(D))`
+2. Reconstruct the root: `computed_root = H(H(A,B), H(C,D))` -> is `computed_root == R`?
+3. Verify: `if (computed_root != R) { err }`
+
+<ArticleSection name="Merkle Trees on Solana" id="merkle-tress-on-solana" level="h2" />
+
+Solana's account-based architecture presents unique challenges for storing large datasets on-chain. Every piece of data requires rent payments, and storing millions of user records directly on-chain becomes prohibitively expensive. 
+
+Merkle trees solve this fundamental scalability problem through compressed accounts: nstead of creating individual accounts for each piece of data, compressed accounts store thousands or millions of data elements as leaves in a Merkle tree, with only the root hash and tree structure living on-chain.
```

Comment:
```suggestion
Merkle trees solve this fundamental scalability problem through compressed accounts: Instead of creating individual accounts for each piece of data, compressed accounts store thousands or millions of data elements as leaves in a Merkle tree, with only the root hash and tree structure living on-chain.
```